sw/device/tests/rom_ext/rom_ext_parser_unittest.cc

```diff
@@ -0,0 +1,185 @@
+#include "sw/device/rom_exts/manifest.h"
+
+namespace rom_ext_parser_unittest {
+namespace {
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::ContainerEq;
+using testing::Each;
+using testing::Eq;
+using testing::Test;
+
+class ParserTest : public Test, public MmioTest {
+ protected:
+  mmio_region_t base_addr_ = {
+      .mock = dev().region(),
+  };
+};
+
+class IdentifierGetTest : public ParserTest {};
+
+TEST_F(IdentifierGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_MANIFEST_IDENTIFIER_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_identifier_get(base_addr_), 0xa5a5a5a5);
+}
+
+class SignatureGetTest : public ParserTest {
+ protected:
+  SignatureGetTest() {
+    for (uint32_t i = 0; i < ROM_EXT_IMAGE_SIGNATURE_SIZE_WORDS; ++i) {
+      src_.data[i] = std::uniform_int_distribution<uint32_t>()(gen_);
+    }
+
+    base_addr_.base = &src_.data[0];
+  }
+
+  rom_ext_signature_t src_;
+
+ private:
+  std::random_device rd;
+  std::mt19937 gen_ = std::mt19937(rd());
+};
+
+TEST_F(SignatureGetTest, Success) {
+  rom_ext_signature_t dst;
+
+  EXPECT_EQ(rom_ext_signature_get(base_addr_, &dst), true);
+  EXPECT_THAT(src_.data, ContainerEq(dst.data));
+}
+
+class ImageLengthGetTest : public ParserTest {};
+
+TEST_F(ImageLengthGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_LENGTH_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_image_len_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ImageVersionGetTest : public ParserTest {};
+
+TEST_F(ImageVersionGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_VERSION_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_version_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ImageTimestampGetTest : public ParserTest {};
+
+TEST_F(ImageTimestampGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_TIMESTAMP_OFFSET, 0xcdcdcdcd);
+  EXPECT_READ32(ROM_EXT_IMAGE_TIMESTAMP_OFFSET + sizeof(uint32_t), 0xabababab);
+
+  uint64_t result = rom_ext_timestamp_get(base_addr_);
+  EXPECT_EQ(result, 0xababababcdcdcdcd);
+}
+
+class AlgorithmIdGetTest : public ParserTest {};
+
+TEST_F(AlgorithmIdGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_SIGNATURE_ALGORITHM_IDENTIFIER_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_algorithm_id_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ExponentGetTest : public ParserTest {};
+
+TEST_F(ExponentGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_EXPONENT_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_exponent_get(base_addr_), 0xa5a5a5a5);
+}
+
+class UsageConstraintsGetTest : public ParserTest {};
+
+TEST_F(UsageConstraintsGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_USAGE_CONSTRAINTS_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_usage_constraints_get(base_addr_), 0xa5a5a5a5);
+}
+
+class PeripheralLockdownInfoGetTest : public ParserTest {
+ protected:
+  PeripheralLockdownInfoGetTest() {
+    for (uint32_t i = 0; i < ROM_EXT_PERIPHERAL_LOCKDOWN_INFO_SIZE_WORDS; ++i) {
+      src_.data[i] = std::uniform_int_distribution<uint32_t>()(gen_);
+    }
+
+    base_addr_.base = &src_.data[0];
+  }
```

Comment:
Consider factoring this code out, since it's repeated three times.

Reply:
It is gone completely now.

Done.